### PR TITLE
py-fenics-basix: update for main and future 0.8.0

### PR DESCRIPTION
**Pull request by chrisrichardson**
Adds new dependency on `py-scikit-build-core`

### CODE REVIEW

**Review by mscroggs — APPROVED** [review]

**Comment by chrisrichardson** [comment]
Waiting for PR #40779

**Review by tldahlgren — COMMENTED** [review]
Hmmm.  v0.8 hasn't been released.  I don't know what the policy is on adding dependencies in this case.  Do you @adamjstewart ?

**Comment by garth-wells** [comment]
> Hmmm. v0.8 hasn't been released. I don't know what the policy is on adding dependencies in this case. Do you @adamjstewart ?

PR title is mis-leading; update is for `main`, and will also work for a future 0.8 release.

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py [hunk 1]
```diff
@@ -35,6 +35,7 @@ class PyFenicsBasix(PythonPackage):
     depends_on("py-numpy@1.21:", type=("build", "run"))
     depends_on("py-pybind11@2.9.1:", when="@:0.7", type="build")
     depends_on("py-nanobind@1.5.1:", when="@0.8:", type="build")
+    depends_on("py-scikit-build-core+pyproject@0.5.0:", when="@0.8:", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-basix/package.py]:
```suggestion
    depends_on("py-nanobind@1.6:", when="@0.8:", type="build")
    depends_on("py-scikit-build-core+pyproject", when="@0.8:", type="build")
```
Based on `pyproject.toml`, unless I'm missing something.

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py [hunk 2]
```diff
@@ -35,6 +35,7 @@ class PyFenicsBasix(PythonPackage):
     depends_on("py-numpy@1.21:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-fenics-basix/package.py:35]: setuptools is no longer required

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py:35]: CMake 3.19+ required now

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py (hunk 1 above)
Comment (chrisrichardson) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py]: I'm not sure what you mean. Deleting `+pyproject` will cause it to fail. Plus we need a new enough version (0.5.0+)

Comment (ma595) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py]: @adamjstewart Could you provide a little more clarity on the above? Thanks.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py]: I was just using the documented dependencies from main, you may want to update this in `pyproject.toml` if those are indeed required.

Comment (garth-wells) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py]: @adamjstewart I've fixed the `pyproject.toml` file in Basix, adding `pyproject` as a build dependency, see PR https://github.com/FEniCS/basix/pull/741 and the line https://github.com/FEniCS/basix/blob/d11d807fab6341b9dcec57d8b89f201d9c6e8636/python/pyproject.toml#L2.

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py (hunk 2 above)
Comment (garth-wells) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py:35]: @ma595 setuptools isn't required for `0@.8:`, but it is required for earlier versions.

Also with CMake: the more recent version CMake is only for `0@.8:`.

**Review by garth-wells — APPROVED** [review]

**Comment by garth-wells** [comment]
@adamjstewart I think this is good now. It's been tested at https://github.com/FEniCS/basix/actions/workflows/spack.yml for `main` and the latest release. And the required dependency update https://github.com/spack/spack/pull/40779 has been merged.

File: var/spack/repos/builtin/packages/py-fenics-basix/package.py (hunk 1 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py]: Thanks, I updated my suggested changes to match the current state of your `pyproject.toml`. Feel free to update the versions either here or in `pyproject.toml`.

Comment (garth-wells) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py]: `scikitb-build-core` version constraint added to updated `pyproject.toml` file, https://github.com/FEniCS/basix/blob/15b915bad133c4c77c9df5c6476506c4960365f5/python/pyproject.toml#L2.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py]: Now nanobind is the last remaining thing that needs to be updated

Comment (garth-wells) [reply var/spack/repos/builtin/packages/py-fenics-basix/package.py]: Oops, overlooked that one. Fixed now.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**